### PR TITLE
Move clients used by admin API to new adminactions

### PR DESCRIPTION
**Pull request by nilsanderselde**
### Which issue this PR addresses:

Fixes https://github.com/Azure/ARO-RP/issues/546, addresses [7844645](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/7844645) and [7862789](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/7862789)

Follow-up to https://github.com/Azure/ARO-RP/pull/898#pullrequestreview-459290004

### What this PR does / why we need it:

`frontend.NewFrontend` has become increasingly unwieldy, with more and more client factories needed to allow for unit testing.

This PR moves client creation into `adminactions.New` (formerly `kubeactions` -- renamed to reflect expanded role). This does away with the need to mock each client in `frontend` unit tests.

### Test plan for issue:

Unit tests for `frontend` have been updated. Unit testing at the `adminactions` level has been added for `ResourcesList` (which was formerly tested at the `frontend` level).

### Is there any documentation that needs to be updated for this PR?

[Geneva Actions Development Guide](https://msazure.visualstudio.com/AzureRedHatOpenShift/_wiki/wikis/ARO.wiki/75878/Geneva-Actions-Development-Guide) in ARO Wiki

### CODE REVIEW

**Comment by ehashman** [comment]
> Is there any documentation that needs to be updated for this PR?
>
> No.

I think https://msazure.visualstudio.com/AzureRedHatOpenShift/_wiki/wikis/ARO.wiki/75878/Geneva-Actions-Development-Guide will need an update on how to do tests?

**Comment by nilsanderselde** [comment]
> > Is there any documentation that needs to be updated for this PR?
> > No.
> 
> I think https://msazure.visualstudio.com/AzureRedHatOpenShift/_wiki/wikis/ARO.wiki/75878/Geneva-Actions-Development-Guide will need an update on how to do tests?

Oh, good call.

**Comment by ehashman** [comment]
@hawkowl @asalkeld @mjudeikis can you PTAL?

File: pkg/frontend/adminactions/serialtool.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	return &serialTool{
+		log: log,
+		oc:  oc,
+
+		accounts:        accounts,
+		virtualMachines: virtualMachines,
+
+		clusterResourceGroup: stringutils.LastTokenByte(oc.Properties.ClusterProfile.ResourceGroupID, '/'),
+	}, nil
+}
```
Comment (asalkeld) [inline pkg/frontend/adminactions/serialtool.go:62]: I think remove this struct and just have one function that creates the client and writes, it's not providing much purpose at this point

File: pkg/frontend/adminactions/resources_list.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/network"
+	"github.com/Azure/ARO-RP/pkg/util/refreshable"
+)
+
+func (a *adminactions) ResourcesList(ctx context.Context, fpAuth refreshable.Authorizer,
+	subscriptionID, clusterRG, vNetID string) ([]byte, error) {
+
+	resourcesClient := features.NewResourcesClient(subscriptionID, fpAuth)
+	vmClient := compute.NewVirtualMachinesClient(subscriptionID, fpAuth)
+	vnetClient := network.NewVirtualNetworksClient(subscriptionID, fpAuth)
```
Comment (asalkeld) [inline pkg/frontend/adminactions/resources_list.go:27]: I think all these clients should go into the adminactions struct and New() should create them, there is not a big overhead and it will make it a lot easy to unit test adminactions

File: pkg/frontend/adminactions/adminactions.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+func (a *adminactions) KubeDelete(oc *api.OpenShiftCluster, groupKind, namespace, name string) error {
+	// TODO log changes
+
+	restConfig, err := restconfig.RestConfig(a.env, oc)
+	if err != nil {
+		return err
+	}
+	dh, err := dynamichelper.New(a.log, restConfig)
```
Comment (asalkeld) [inline pkg/frontend/adminactions/adminactions.go:107]: see: https://msazure.visualstudio.com/AzureRedHatOpenShift/_backlogs/backlog/Azure%20Red%20Hat%20OpenShift/Features/?showParents=true&workitem=7862789 (this should go into the adminactions struct as well - could be a different PR)

File: pkg/frontend/admin_openshiftcluster_resources_list_test.go [hunk 4]
```diff
@@ -115,43 +108,19 @@ func TestAdminListResourcesList(t *testing.T) {
 				subscriptions.EXPECT().Get(gomock.Any(), mockSubID).
 					Return(subscriptionDoc, nil)
 
-				resources.EXPECT().List(gomock.Any(), "resourceGroup eq 'test-cluster'", "", nil).Return([]mgmtfeatures.GenericResourceExpanded{
```
Comment (asalkeld) [inline pkg/frontend/admin_openshiftcluster_resources_list_test.go:118]: these tests need to make it into adminactions/resourceslist_test.go

**Comment by mjudeikis** [comment]
lets not merge this until we do a rollout, please. Not a priority to get in as it does not give any functionality now and just introduces more risk.

File: pkg/frontend/admin_openshiftcluster_kubernetesobjects.go [hunk 5]
```diff
@@ -59,9 +59,9 @@ func (f *frontend) _getAdminKubernetesObjects(ctx context.Context, r *http.Reque
 	}
 
 	if name != "" {
-		return f.kubeActionsFactory(log, f.env).Get(doc.OpenShiftCluster, groupKind, namespace, name)
+		return f.adminActionsFactory(log, f.env).KubeGet(doc.OpenShiftCluster, groupKind, namespace, name)
```
Comment (kwoodson) [inline pkg/frontend/admin_openshiftcluster_kubernetesobjects.go:62]: I prefer the old style names.  `Get|List|Delete|CreateOrUpdate`.

File: pkg/frontend/adminactions/adminactions.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	"github.com/Azure/ARO-RP/pkg/dynamichelper"
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/compute"
+	"github.com/Azure/ARO-RP/pkg/util/refreshable"
+	"github.com/Azure/ARO-RP/pkg/util/restconfig"
+)
+
+// Interface for adminactions
+type Interface interface {
+	KubeGet(oc *api.OpenShiftCluster, groupKind, namespace, name string) ([]byte, error)
```
Comment (kwoodson) [inline pkg/frontend/adminactions/adminactions.go:25]: I prefer not having `kube` in the names here.

**Comment by kwoodson** [comment]
Looking good @nilsanderselde!

File: pkg/frontend/adminactions/resources_list.go (hunk 2 above)
Comment (nilsanderselde) [reply pkg/frontend/adminactions/resources_list.go:27]: That was in the previous refactoring PR, but I thought we didn't want it because of overhead (make them all every time for every action). But if that's not a big deal, then I agree it would be cleaner to make them in one place.

File: pkg/frontend/adminactions/serialtool.go (hunk 1 above)
Comment (nilsanderselde) [reply pkg/frontend/adminactions/serialtool.go:62]: Good point

File: cmd/aro/rp.go [hunk 7]
```diff
@@ -96,8 +93,7 @@ func rp(ctx context.Context, log *logrus.Entry) error {
 		return err
 	}
 
-	// TODO(mj): We need to fix this argument chain
```
Comment (nilsanderselde) [inline cmd/aro/rp.go:99]: ✔

File: pkg/frontend/admin_openshiftcluster_mustgather.go [hunk 8]
```diff
@@ -40,8 +40,15 @@ func (f *frontend) _postAdminOpenShiftClusterMustGather(ctx context.Context, w h
 		return err
 	}
 
-	w.Header().Add("Content-Type", "application/gzip")
-	w.Header().Add("Content-Disposition", `attachment; filename="must-gather.tgz"`)
```
Comment (nilsanderselde) [inline pkg/frontend/admin_openshiftcluster_mustgather.go:44]: Moved to right before `io.Copy` to be consistent with https://github.com/Azure/ARO-RP/pull/886#discussion_r463724754

File: pkg/frontend/adminactions/mustgather.go [hunk 9]
```diff
@@ -151,24 +141,27 @@ func (ka *kubeactions) MustGather(ctx context.Context, oc *api.OpenShiftCluster,
...
-	rc, err := portforward.ExecStdout(ctx, ka.log, ka.env, oc, pod.Namespace, pod.Name, "copy", []string{"tar", "cz", "/must-gather"})
+	a.log.Info("must-gather pod running")
+	rc, err := portforward.ExecStdout(ctx, a.log, a.env, a.oc, pod.Namespace, pod.Name, "copy", []string{"tar", "cz", "/must-gather"})
 	if err != nil {
 		return err
 	}
 	defer rc.Close()
 
+	w.Header().Add("Content-Type", "application/gzip")
+	w.Header().Add("Content-Disposition", `attachment; filename="must-gather.tgz"`)
```
Comment (nilsanderselde) [inline pkg/frontend/adminactions/mustgather.go:158]: Moved here to be consistent with https://github.com/Azure/ARO-RP/pull/886#discussion_r463724754

File: pkg/frontend/admin_openshiftcluster_kubernetesobjects_test.go [hunk 10]
```diff
@@ -110,21 +125,34 @@ func TestAdminKubernetesObjectsGetAndDelete(t *testing.T) {
...
+			mocks: func(tt *test, a *mock_adminactions.MockInterface, oc *mockdatabase.MockOpenShiftClusters, s *mockdatabase.MockSubscriptions) {
 				clusterDoc := &api.OpenShiftClusterDocument{
+					Key: tt.resourceID,
 					OpenShiftCluster: &api.OpenShiftCluster{
 						ID:   "fakeClusterID",
 						Name: "resourceName",
 						Type: "Microsoft.RedHatOpenShift/openshiftClusters",
 					},
 				}
+				subscriptionDoc := &api.SubscriptionDocument{
```
Comment (nilsanderselde) [inline pkg/frontend/admin_openshiftcluster_kubernetesobjects_test.go:137]: This is needed now since a standard set of clients is created for all `adminactions`

File: pkg/frontend/admin_openshiftcluster_kubernetesobjects_test.go [hunk 11]
```diff
@@ -137,7 +165,7 @@ func TestAdminKubernetesObjectsGetAndDelete(t *testing.T) {
 			resourceID:   fmt.Sprintf("/subscriptions/%s/resourcegroups/resourceGroup/providers/Microsoft.RedHatOpenShift/openShiftClusters/resourceName", mockSubID),
 			objNamespace: "openshift-project",
 			objName:      "config",
-			mocks: func(tt *test, openshiftClusters *mock_database.MockOpenShiftClusters, kactions *mock_kubeactions.MockInterface) {
+			mocks: func(tt *test, a *mock_adminactions.MockInterface, oc *mockdatabase.MockOpenShiftClusters, s *mockdatabase.MockSubscriptions) {
```
Comment (nilsanderselde) [inline pkg/frontend/admin_openshiftcluster_kubernetesobjects_test.go:168]: I shortened and standardized these names and standardized the order across the tests for readability.

File: pkg/frontend/admin_openshiftcluster_list_test.go [hunk 12]
```diff
@@ -75,7 +75,7 @@ func TestAdminListOpenShiftCluster(t *testing.T) {
 	for _, tt := range []*test{
 		{
 			name: "clusters exists in db",
-			mocks: func(controller *gomock.Controller, openshiftClusters *mock_database.MockOpenShiftClusters, enricher *mock_clusterdata.MockOpenShiftClusterEnricher, cipher *mock_encryption.MockCipher) {
+			mocks: func(controller *gomock.Controller, oc *mock_database.MockOpenShiftClusters, enricher *mock_clusterdata.MockOpenShiftClusterEnricher, cipher *mock_encryption.MockCipher) {
```
Comment (nilsanderselde) [inline pkg/frontend/admin_openshiftcluster_list_test.go:78]: Shortened this name to be consistent with the other tests

File: pkg/frontend/admin_openshiftcluster_resources_list.go [hunk 13]
```diff
@@ -36,22 +28,25 @@ func (f *frontend) listAdminOpenShiftClusterResources(w http.ResponseWriter, r *
...
 	vars := mux.Vars(r)
 	resourceID := strings.TrimPrefix(r.URL.Path, "/admin")
 
 	doc, err := f.db.OpenShiftClusters.Get(ctx, resourceID)
 	switch {
 	case cosmosdb.IsErrorStatusCode(err, http.StatusNotFound):
-		return nil, api.NewCloudError(http.StatusNotFound, api.CloudErrorCodeResourceNotFound, "", "The Resource '%s/%s' under resource group '%s' was not found.", vars["resourceType"], vars["resourceName"], vars["resourceGroupName"])
+		return nil, api.NewCloudError(http.StatusNotFound, api.CloudErrorCodeResourceNotFound, "",
+			"The Resource '%s/%s' under resource group '%s' was not found.",
+			vars["resourceType"], vars["resourceName"], vars["resourceGroupName"])
```
Comment (nilsanderselde) [inline pkg/frontend/admin_openshiftcluster_resources_list.go:49]: _this line was too long_

File: pkg/frontend/admin_openshiftcluster_resources_list_test.go (hunk 4 above)
Comment (nilsanderselde) [reply pkg/frontend/admin_openshiftcluster_resources_list_test.go:118]: I will work on this next.

File: pkg/frontend/adminactions/adminactions.go [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+		log: log,
+		env: env,
+		oc:  oc,
+
+		dh:              dh,
+		k8sClient:       k8sClient,
+		resourcesClient: features.NewResourcesClient(subscriptionDoc.ID, fpAuth),
+		vmClient:        compute.NewVirtualMachinesClient(subscriptionDoc.ID, fpAuth),
+		vNetClient:      network.NewVirtualNetworksClient(subscriptionDoc.ID, fpAuth),
+		accountsClient:  storage.NewAccountsClient(subscriptionDoc.ID, fpAuth),
```
Comment (nilsanderselde) [inline pkg/frontend/adminactions/adminactions.go:89]: These now live here. Much simpler.

File: pkg/frontend/adminactions/mustgather.go [hunk 15]
```diff
@@ -15,24 +16,13 @@ import (
...
-	restconfig, err := restconfig.RestConfig(ka.env, oc)
-	if err != nil {
-		return err
-	}
-
-	cli, err := kubernetes.NewForConfig(restconfig)
-	if err != nil {
-		return err
-	}
+func (a *adminactions) MustGather(ctx context.Context, w http.ResponseWriter) error {
```
Comment (nilsanderselde) [inline pkg/frontend/adminactions/mustgather.go:23]: > `http.ResponseWriter`

This is what it was being passed, anyway. It needs to be changed here so headers can be added in this function, which is a better location to do that.

**Comment by nilsanderselde** [comment]
> lets not merge this until we do a rollout, please. Not a priority to get in as it does not give any functionality now and just introduces more risk.

@mjudeikis I added the `hold` label.

File: pkg/frontend/admin_openshiftcluster_resources_list_test.go (hunk 4 above)
Comment (nilsanderselde) [reply pkg/frontend/admin_openshiftcluster_resources_list_test.go:118]: Added file `pkg/frontend/adminactions/resources_list_test.go`

File: pkg/frontend/adminactions/resources_list.go (hunk 2 above)
Comment (nilsanderselde) [reply pkg/frontend/adminactions/resources_list.go]: @asalkeld Done

File: pkg/frontend/adminactions/adminactions.go (hunk 6 above)
Comment (asalkeld) [reply pkg/frontend/adminactions/adminactions.go]: Yes, but now admin.Get() is what exactly? the cluster/a resource or a kube object. i think it is needed to know what it is doing

File: pkg/frontend/adminactions/serialtool.go [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	blobService := azstorage.NewAccountSASClient(
+		"cluster"+a.oc.Properties.StorageSuffix, v, azure.PublicCloud).GetBlobService()
+
+	c := blobService.GetContainerReference(parts[1])
+
+	b := c.GetBlobReference(parts[2])
+
+	rc, err := b.Get(nil)
```
[the remark this answers is not in the document]
Comment (nilsanderselde) [reply pkg/frontend/adminactions/serialtool.go:70]: Yeah - this function will need refactoring.

**Comment by nilsanderselde** [comment]
@asalkeld Made those changes

File: pkg/frontend/adminactions/resources_list.go [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+	mgmtcompute "github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-03-01/compute"
+	"github.com/Azure/go-autorest/autorest/azure"
+
+	"github.com/Azure/ARO-RP/pkg/util/arm"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient"
+	"github.com/Azure/ARO-RP/pkg/util/stringutils"
+	"github.com/Azure/ARO-RP/pkg/util/subnet"
+)
+
+func (a *adminactions) ResourcesList(ctx context.Context) ([]byte, error) {
```
Comment (asalkeld) [inline pkg/frontend/adminactions/resources_list.go:20]: space here.. we really need a linter for this..

**Review by asalkeld — COMMENTED** [review]
I am happy with this (just kill the spaces on the first line of functions)

File: pkg/frontend/admin_openshiftcluster_resources_list.go [hunk 18]
```diff
@@ -36,22 +28,25 @@ func (f *frontend) listAdminOpenShiftClusterResources(w http.ResponseWriter, r *
...
 	if err == nil {
 		b, err = adminJmespathFilter(b, jpath)
 	}
 
 	adminReply(log, w, nil, b, err)
 }
 
-func (f *frontend) _listAdminOpenShiftClusterResources(ctx context.Context, r *http.Request) ([]byte, error) {
+func (f *frontend) _listAdminOpenShiftClusterResources(
+	ctx context.Context, r *http.Request, log *logrus.Entry) ([]byte, error) {
```
Comment (jim-minter) [inline pkg/frontend/admin_openshiftcluster_resources_list.go:40]: @nilsanderselde FWIW I'm not wild about all the newlines. From https://golang.org/doc/effective_go.html : "Go has no line length limit. Don't worry about overflowing a punched card."  In general this codebase aims to wrap full-line comments at 80 and doesn't care too much about long code lines unless they get really wild.

**Comment by jim-minter** [comment]
@nilsanderselde (cc @asalkeld) this looks excellent and I'm happy to merge it if you are.  I pushed two extra commits, one which is a nit to move the pod ready check to pkg/util/ready; the other one is more important - it should avoid calling out to the cluster to populate the APIs for the dynamic helper unless we are actually going to do a k8s call to the cluster (saves a bunch of API calls when we're doing an Azure-level geneva action).

File: pkg/frontend/adminactions/resources_list.go (hunk 17 above)
Comment (nilsanderselde) [reply pkg/frontend/adminactions/resources_list.go:20]: that would be nice
